### ISSUE TITLE
Rewriting Lebab as a Babel preset

### ISSUE THREAD

**Issue by abiduzz420** [babel/babel#7296]
I have been looking into lebab project and learning to write codemods from past few days. To me the idea of converting esOld to esNew (5to6) is very intriguing and it would be a massive boost for programmers to perform painless migrations with minimum human intervention. I have gone through this issue [@lebab/lebab#138](https://github.com/lebab/lebab/issues/138) and read the [proposal](https://github.com/babel/babel/wiki/Google-Summer-of-Code-2017-Proposals#rewrite-lebab-as-a-babel-preset) of re-writing lebab as a babel preset. I am also looking into `babel-codemod` by @eventualbuddha, how React is solving the issue of breaking API changes and tinkered with different tools out there(js-codeshift, cpojer/js-codemod etc). Along the way, I have made few conclusions and would like to know if I am thinking on the correct lines.

- Our aim is to write babel transformations taking inspiration from the lebab project.
- We are using `recast` as parser since it helps create a shadow AST which lets us perform tree diffing, which is later used to generate source code which adheres to the input code style.
- We have also decided to use `recast` as code generator instead of `babel-generator`, since it preserves the code style and in this case, it is outputted to `src` dir. We are not using `babel-generator` as it removes white spaces, changes code style and hence the diff becomes unreadable.
- We need to write bunch of babel-plugins/codemods which are to be grouped together as a preset.
- Finally the idea in action would look like below:
1. We configure babel in `.babelrc` as
```
{
  "parserOpts": {
    "parser": "recast"
  },
  "generatorOpts": {
    "generator": "recast"
  }
}
```
2. Run the command `babel src -d src --presets=lebab`

To summarize, this is what it looks like:
1. Parser: recast
2. Transformer: Babel
3. Code generator: recast
4. Lebab consisting of set of babel plugins and hence used as babel preset. 
5. Modified code is outputted to `src` directory itself.

cc: @hzoo @loganfsmyth @existentialism @drewml @sven @benjamn @nene

**Comment by hzoo** [comment]
Thanks for the issue, glad you are interested in the project!

The idea was actually *not* to remove Lebab as a tool (which is why the issue is still in that repo) but to use the same CLI tool, but internally it would "use" babel itself. So yes, we would still implement a babel preset but the CLI doesn't have to be with babel unless someone wants to do that.

You shouldn't need to configure it with `parserOpts, recast, presets` etc because it's a different use case, as Lebab already does everything as it should. Ideally you would just use it globally or with npx so there isn't a need to configure it. Unless we want to just use babel-codemod and wrap that to make lebab itself.

**Comment by abiduzz420** [comment]
Hello @hzoo I have been reading about ASTs, codemods and understanding the lebab code base. However, I am not able to figure out my first step to working on this issue. Could you please guide me on what my next step should be ? 

To my understanding I feel I should be refactoring lebab code into plugin based architecture, which could be then used as babel preset. What do you think ?

**Comment by benjamn** [comment]
To be perfectly honest with you, Lebab is @nene's project (with help from @mohebifar), and is not directly affiliated with Babel, so you should probably be asking about this on the Lebab repository (e.g. https://github.com/lebab/lebab/issues/138). Specifically, you might ask if anyone else has done similar work already, study what they did, and ask them lots of questions. In general, I would caution against trying to make progress without first establishing feasibility and enthusiasm with @nene and others who are most familiar with the project. Yes, it's all open source, but unless your ideas are welcome to the maintainer(s) of the project, it will be very hard to deliver meaningful improvements.

**Comment by abiduzz420** [comment]
Thank you @benjamn . That sounds correct. I shall get in touch with lebab project maintainers and take it forward from there :)

**Comment by danez** [comment]
Closing as there is nothing here we can do at the moment and there is an issue in the lebab repo.
